standards/core/rln-contract.md

```diff
@@ -0,0 +1,301 @@
+---
+title: WAKU2-RLN-CONTRACT
+name: Waku2 RLN Contract Specification
+category: Standards Track
+tags: [waku/core-protocol]
+editor: Sergei Tikhomirov <sergei@status.im>
+contributors:
+---
+
+## Abstract
+
+This document describes membership management for the RLN smart contract, in particular:
+- membership-related contract functionality;
+- suggested parameters valued for the initial mainnet deployment;
+- contract governance and upgradability.
+
+This document currently only considers membership-related functionality.
+It might later evolve into a full-fledged contract specification.
+
+RLN is only deployed on Sepolia testnet ([source code](https://github.com/waku-org/waku-rlnv2-contract/blob/main/src/WakuRlnV2.sol)) as of August 2024.
+This document aims to outline the path to its mainnet deployment.
+
+## Syntax
+
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “NOT RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt).
+
+## Background
+
+Rate-Limiting Nullifier (RLN) is a ZK-based gadget used for privacy-preserving rate limiting in Waku.
+The RLN smart contract is the core element of RLN architecture.
+The smart contract stores the RLN tree that contains all currently existing memberships.
+Users interact with the contract to manage their memberships
+and to get the necessary data for proof generation and verification.
+
+Sending messages is handled by Waku RLN Relay nodes.
+To send a message, the sender MUST prove its validity in terms of RLN.
+An RLN Relay node MUST relay a message unless it is invalid.
+See [17/WAKU2-RLN-RELAY](https://github.com/vacp2p/rfc-index/blob/main/waku/standards/core/17/rln-relay.md) for the full specification of RLN Relay.
+
+## Contract overview
+
+The contract MUST provide the following functionalities:
+- register a membership;
+- extend a membership;
+- withdraw a deposit.
+
+Contract parameters and their RECOMMENDED values for the initial mainnet deployment are as follows:
+
+| Parameter                                   | Symbol    | Value   | Units                              |
+| ------------------------------------------- | --------- | ------- | ---------------------------------- |
+| Epoch length                                | `epoch`   | `10`    | minutes                            |
+| Maximum total rate limit of all memberships | `R_{max}` | `20000` | messages per `epoch`               |
```

Comment:
That's 5Mbps, right? for the network, around 0.6Mbps per shard.

Shall we start higher? 3200 -> 8Mbps for network, around 1Mbps per shard.

Reply:
20 thousand was pulled from an earlier document, I don't have a strong personal opinion here.

However, I vaguely remember some internal research on what the average (median?) connection speeds are, and the conclusion was that in Africa, the slowest continent in that respect, it's 10 (??) Mbps, so if we want Waku to work globally, we should think of this bandwidth cap. Is this true, and is this relevant?

Reply:
I'm changing to 160'000 for now, let's revisit this later if there are any issues.